lisp/ess-r-flymake.el

```diff
@@ -62,24 +64,45 @@ each element is passed as argument to 'lintr::with_defaults'."
 
 (defvar-local ess-r--flymake-proc nil)
 
+(defvar-local ess-r--lintr-file nil
+  "Location of the .lintr file for this buffer.")
+
 (defvar ess-r--flymake-def-linter
   (replace-regexp-in-string
    "[\n\t ]+" " "
-   "esslint <- function(str, linters, cache) {
+   "esslint <- function(str, ...) {
     if (!suppressWarnings(require(lintr, quietly=T))) {
         cat('@@error: @@`lintr` package not installed')
     } else {
         if (packageVersion('lintr') <= '1.0.1') {
             cat('@@error: @@Need `lintr` version > v1.0.1')
         } else {
-            tryCatch(lintr::lint(commandArgs(TRUE), linters = linters, cache = cache),
+            tryCatch(lintr::lint(commandArgs(TRUE), ...),
                     error = function(e) {
                        cat('@@warning: @@', e)
                     })
         }
     }
 };"))
 
+(defun ess-r--find-lintr-file ()
+  "Return the absolute path to the .lintr file.
+Check first the current directory, then the project root, then
+the user's home directory.  Return nil if we couldn't find a .lintr file."
+  (cond (;; current directory
+         (file-readable-p (expand-file-name ".lintr" default-directory))
+         (expand-file-name ".lintr" default-directory))
+        ;; Project root
+        ((and (fboundp 'ess-r-package-project)
+              (ess-r-package-project)) (expand-file-name ".lintr" (cdr (ess-r-package-project))))
+        ((and (fboundp 'project-roots)  ; checking that project-roots is defined can be removed when drop Emacs 24
+              (project-current)
+              (project-roots (project-current)))
+         (expand-file-name ".lintr" (car (project-roots (project-current)))))
```

Comment:
It should loop over roots and check if file exists I believe. Apparently there are systems in the wild with multiple roots.

Reply:
That's strange. Anyway, I just pushed a new commit that does that.

Reply:
There is another issue with project-roots. It will not work on our projects of the form `(R-project . "path"). We need to define our own `project-roots` method. 

I am having doubts about project.el :/ There are still too many unpolished edges.

Reply:
OOP doesn't automatically mean that I have to copy-paste methods every time I define a new "class". There should be sane default methods.

Reply:
What reason is there to assume that the cdr of a project instance is a list of strings?

Reply:
@vspinu FWIW I don't think that affects this since we check if `ess-r-package-project` returns non-nil before checking project.el stuff.

Reply:
The problem with such a loose-typed approach is that an author of any backend that doesn't conform to this scheme will have a harder time debugging why `project-roots` doesn't work properly.

Right now, the answer to that question is simple: "you haven't defined it for your project".

> Wouldn't (proj-type root1 root2 ...) as project instance be enough for pretty much everything?

Well, there are also "external roots", for one thing.